File: SwiftUI/EdgeStudio/Models/MulticastConfig.swift
import Foundation

/// Settings for the reliable UDP multicast transport (beta, Ditto SDK 5.1.0 —
/// `DittoPeerToPeer.multicastBeta`).
///
/// Persisted per database (four `databaseConfigs` columns, SQLCipher schema v6).
/// Defaults match the SDK defaults, so a config equal to `MulticastConfig()` is a
/// no-op. Default is **disabled** — unlike the other p2p transports, multicast
/// requires all peers on the same L2 segment.
///
/// Field-proven rule (from the Zava Retail demo, verified on-device): port 0 is
/// rejected because the SDK treats it as "pick any port", which silently breaks
/// group rendezvous between peers.
struct MulticastConfig: Equatable, Codable, Sendable {
    var isEnabled = false
    var groupAddress: String = MulticastConfig.defaultGroupAddress
    var port: Int = MulticastConfig.defaultPort
    var interfaceName: String?

    static let defaultGroupAddress = "224.1.2.3"
    static let defaultPort = 6003

    /// IPv4 class-D dotted-quad: four octets 0...255, the first in 224...239.
    ///
    /// Strict parsing on purpose: `Int()` alone accepts `+224`, leading zeros,
    /// and non-ASCII digits, and `split(separator:)` omits empty fields so
    /// `224..1.2.3` would collapse to three parts and could slip through. Each
    /// octet must be plain ASCII digits, no leading zeros beyond "0" itself.
    static func isValidGroupAddress(_ address: String) -> Bool {
        let parts = address
            .trimmingCharacters(in: .whitespaces)
            .split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 4 else { return false }
        let octets = parts.compactMap { part -> Int? in
            guard !part.isEmpty,
                  part.allSatisfy({ $0.isASCII && $0.isNumber }),
                  part.count == 1 || !part.hasPrefix("0") else { return nil }
            return Int(part)
        }
        guard octets.count == 4, octets.allSatisfy({ (0 ... 255).contains($0) }) else {
            return false
        }
        return (224 ... 239).contains(octets[0])
    }

    /// Parses `text` as a UDP port; nil unless a whole number in 1...65535.
    static func parsePort(_ text: String) -> Int? {
        guard let port = Int(text.trimmingCharacters(in: .whitespaces)),
              (1 ... 65535).contains(port) else { return nil }
        return port
    }

    /// The port in the SDK's `UInt16` shape. Clamping is unreachable once input has
    /// passed `parsePort`, so this never silently changes a validated value.
    var sdkPort: UInt16 {
        UInt16(clamping: port)
    }
}

File: android/app/src/main/java/com/costoda/dittoedgestudio/domain/model/MulticastConfig.kt
package com.costoda.dittoedgestudio.domain.model

/**
 * Settings for the reliable UDP multicast transport (beta, Ditto SDK 5.1.0 —
 * `DittoTransportConfig.PeerToPeer.MulticastBeta`).
 *
 * Persisted per database (four `databaseConfigs` columns, Room v7). Defaults match
 * the SDK defaults, so a config equal to `MulticastConfig()` is a no-op.
 *
 * Field-proven rules (from the Zava Retail demo, verified on-device with 5.1.0):
 * - The SDK **defers multicast changes while sync is active** — callers must stop
 *   sync, apply, then restart sync (the existing transport-apply path already does).
 * - Port 0 is rejected: the SDK treats it as "pick any port", which silently breaks
 *   group rendezvous between peers.
 */
data class MulticastConfig(
    val enabled: Boolean = false,
    val groupAddress: String = DEFAULT_GROUP_ADDRESS,
    val port: Int = DEFAULT_PORT,
    val interfaceName: String? = null,
) {
    companion object {
        const val DEFAULT_GROUP_ADDRESS = "224.1.2.3"
        const val DEFAULT_PORT = 6003

        /**
         * IPv4 class-D dotted-quad: four octets 0..255, first in 224..239.
         *
         * Strict octet parsing, mirroring SwiftUI's validator and the SDK's
         * Rust `Ipv4Addr::parse` (the validator the SDK runs at
         * `sync.start()`): Kotlin's `toIntOrNull` accepts leading zeros,
         * a leading `+`, and non-ASCII digits, all of which the SDK rejects.
         * A value that only fails at sync start is persisted first and then
         * fails on every open — it must be rejected here, at the boundary.
         * Note `parsePort` is intentionally NOT strictened: the port crosses
         * the FFI as an integer, so "+6003"/"06003" parse to the same value.
         */
        fun isValidGroupAddress(address: String): Boolean {
            val parts = address.trim().split(".")
            if (parts.size != 4) return false
            val octets = parts.map { part ->
                if (part.isEmpty() || part.any { it !in '0'..'9' }) return false
                if (part.length > 1 && part.startsWith("0")) return false
                part.toIntOrNull()?.takeIf { it in 0..255 } ?: return false
            }
            return octets.first() in 224..239
        }

        /** Parses [text] as a UDP port; null unless a whole number in 1..65535. */
        fun parsePort(text: String): Int? =
            text.trim().toIntOrNull()?.takeIf { it in 1..65535 }
    }
}
